src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+internal sealed partial class DelegateCacheRewriter
+{
+    private readonly SyntheticBoundNodeFactory _factory;
+    private readonly MethodSymbol _topLevelMethod;
+    private readonly int _topLevelMethodOrdinal;
+
+    private Stack<(int LocalFunctionOrdinal, DelegateCacheContainer CacheContainer)>? _genericCacheContainers;
+
+    internal DelegateCacheRewriter(SyntheticBoundNodeFactory factory, int topLevelMethodOrdinal)
+    {
+        Debug.Assert(factory.TopLevelMethod is { });
+
+        _factory = factory;
+        _topLevelMethod = factory.TopLevelMethod;
+        _topLevelMethodOrdinal = topLevelMethodOrdinal;
+    }
+
+    internal static bool CanRewrite(CSharpCompilation compilation, MethodSymbol topLevelMethod, bool inExpressionLambda, BoundConversion boundConversion, MethodSymbol targetMethod)
+        => targetMethod.IsStatic
+        && !boundConversion.IsExtensionMethod
+        && !inExpressionLambda // The tree structure / meaning for expression trees should remain untouched.
+        && topLevelMethod.MethodKind != MethodKind.StaticConstructor // Avoid caching twice if people do it manually.
+        && compilation.IsStaticMethodGroupDelegateCacheEnabled
+        ;
+
+    internal BoundExpression Rewrite(int localFunctionOrdinal, SyntaxNode syntax, BoundExpression receiver, MethodSymbol targetMethod, TypeSymbol delegateType)
+    {
+        Debug.Assert(delegateType.IsDelegateType());
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(localFunctionOrdinal, delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(null, cacheField);
+        var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, receiver, targetMethod, isExtensionMethod: false, type: delegateType);
```

Comment:
>new BoundDelegateCreationExpression(syntax, receiver, targetMethod, isExtensionMethod: false, type: delegateType) [](http://example.com/codeflow?start=36&length=113)

This looks like a code duplication. Consider passing the node that allocates new instance of a delegate as a parameter.

Reply:
Good point, let me see.